### PR TITLE
BUG: Handle unit count when converting from `np.datetime64`

### LINKED ISSUE #39977: BUG: numpy.datetime64 unit counts ignored
- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the latest version of pandas.

- [X] (optional) I have confirmed this bug exists on the master branch of pandas.

---

```python
import numpy as np
import pandas as pd
pd.Timestamp(np.datetime64('2000-01-01', '10ns')) == pd.Timestamp('2000-01-01')
```

Should be ```True```, is instead ```False```. The problem is that the unit count of 10 is ignored. Compare:

```python
pd.Timestamp(np.datetime64('2000-01-01', '10ns')).asm8.item()
pd.Timestamp(np.datetime64('2000-01-01', '1ns')).asm8.item()                                        
```

The bug occurs in two places, both in `_libs/tslibs/conversion.pyx`. Both `get_datetime64_nanos` and `ensure_datetime64ns` forget to check the count. I don't know a Cython API to get the count, though it is accessible from `numpy.datetime_data`.

### PR DESCRIPTION
**Pull request by rob-sil**
- [x] closes #39977
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/v3.0.0.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/_libs/tslibs/np_datetime.pyx [hunk 1]
```diff
@@ -78,6 +78,13 @@ cdef NPY_DATETIMEUNIT get_datetime64_unit(object obj) noexcept nogil:
     return <NPY_DATETIMEUNIT>(<PyDatetimeScalarObject*>obj).obmeta.base
 
 
+cdef int get_datetime64_unit_num(object obj) noexcept nogil:
+    """
+    returns the number of units of the dtype for a numpy datetime64 object.
+    """
+    return (<PyDatetimeScalarObject*>obj).obmeta.num
```
Comment (WillAyd) [inline pandas/_libs/tslibs/np_datetime.pyx:85]: What does number of units mean exactly? And is there a public way to get this - I am wary of relying up internals llike this

**Comment by jbrockmendel** [comment]
I think this case is rare enough I’d rather not support it at all than support is inconsistently across the codebase.  Iirc @seberg was thinking of deprecating it in numpy too

**Comment by seberg** [comment]
Yeah, I suspect very few use that really and unless someone would shout, I would be fine with deprecating it.  I would also suspect NumPy also has bugs around it due to it having at least almost no users in practice.

**Comment by rob-sil** [comment]
Any interest in adding a warning or putting a note in the docs? Otherwise, I'll close this PR.

**Comment by mroeschke** [comment]
Looks like there not much appetite for this change so closing

**Pull request closed without merging**